---
title: "projectPBL_part3_4_5"
author: "Yin Lei & Lisa Spindler"
date: "2022-12-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
```

# TODO: go through chapter 8 for analysis!

## 1: Read Data

Description       | Temperature | Time(s)
----------------- | ----------- | -------
Reference         | 25°C        | 0 min
Mild heat shock   | 37°C        | 10 min, 30 min
Severe heat shock | 42°C        | 10 min, 30 min

Yeast types:
* Reference/Wildtype
* Knockdown: HSF1
* Knockout: MSN2/4
* Knockdown & Knockout: HSF1 and MSN2/4 (double)

$\to$ 3 replicates of each condition


```{r}
library(tidyverse)
complex_yeast_heatshock <- read_tsv("complex_yeast_heatshock.tsv")
sample_mapping <- read_tsv("sample_mapping.tsv")
flat_file <- read_tsv("yeastract2019-flat-file.tsv", col_names = F)
names(flat_file) <- c("Regulator_systematic", "Regulator_standard",
                      "Target_systematic", "Target_standard",
                      "Id", "Date", "Environment", "Regulation", "Type", "Method")
```

## Get to Know the Data

First of all, we looked at the data including the mapping in order to get a basic understanding of the data.

### Difference between KO and KD 

"The **key difference** between gene knockout and knockdown is that the **gene knockout is a technique where the gene of interest is completely removed (inoperative state)** to study of functions of the gene while **gene knockdown is another technique where the gene of interest is silenced** to investigate the role of the particular gene in a biological system." - [Difference Between Gene Knockout and Knockdown](https://www.differencebetween.com/difference-between-gene-knockout-and-knockdown/)

### Observations on Our Data

* MSN2/4 is always KO
* HSF1 is always KD
* Wildtype $\to$ without any KO/KD
* Double $\to$ MSN2/4 & HSF1 both KO/KD
* Temperature in °C 
* For each condition and strain: 3 replicates
* 25°C / 37°C / 42°C after certain times (0 min for 25°C; 10/30 min for heatshock)
* Gene expression data for each gene
* Flat file for interactions

### Get an Impression of HSF1 and MSN2/4

#### HSF1

"Trimeric heat shock transcription factor; activates multiple genes in response to highly diverse stresses and transient intracellular acidification; recognizes variable heat shock elements (HSEs) consisting of inverted NGAAN repeats; monitors translational status of cell through an RQC (Ribosomal Quality Control)-mediated translation-stress signal; involved in diauxic shift; posttranslationally regulated; human homolog HSF1 with linker region mutations can complement yeast hsf1 mutant" -- [Description of HSF1 / YGL073W in SGD](https://www.yeastgenome.org/locus/S000003041)

#### MSN2 

"Stress-responsive transcriptional activator; activated in stochastic pulses of nuclear localization in response to various stress conditions; binds DNA at stress response elements of responsive genes; light sensing pathway component that accumulates in the nucleus in response to blue light; relative distribution to nucleus increases upon DNA replication stress" -- [Description of MSN2 / YMR037C in SGD](https://www.yeastgenome.org/locus/S000004640)

#### MSN4

"Stress-responsive transcriptional activator; activated in stochastic pulses of nuclear localization in response to various stress conditions; binds DNA at stress response elements of responsive genes, inducing gene expression; involved in diauxic shift" -- [Description of MSN4 / YKL062W in SGD](https://www.yeastgenome.org/locus/S000001545)

#### Diauxic Shift - Short Explanation

"The switch from rapid fermentative growth in the presence of a rich carbon source to slower exponential growth by aerobic respiration using ethanol once the preferred carbon source has been exhausted." -- [Description of diauxic shift in SGD](https://www.yeastgenome.org/keyword/diauxic_shift)

---

"A diauxic shift occurs when a microorganism is grown in a batch culture with two substrates. Rather than metabolizing the two available sugars simultaneously, microbial cells commonly consume them in a sequential pattern, resulting in two separate growth phases.

When growing Saccharomyces cerevisiae with glucose and plenty of aeration, the diauxic growth pattern is commonly observed in a batch culture. During the first growth phase, when there is plenty of glucose and oxygen available, the yeast cells prefer glucose fermentation to aerobic respiration (crabtree effect). After glucose is depleted, the fermentation product ethanol is oxidised in a noticeably slower second growth phase, if oxygen is available." -- [Diauxic shift on Labster Theory](https://theory.labster.com/diauxic-shift/)

![Typical growth curve of yeast in high glucose substrate.\nImage Source: https://theory.labster.com/diauxic-shift/](https://s3-us-west-2.amazonaws.com/labster/wiki/media/Growth_phase1.jpg)

### Interaction Overview with HSF1 & MSN2/4

filter flat file for interactions with HSF1 and MSN2/4 and create a graph

Look at Felix's code in moodle

### Analysis: Heatmap 

To get an overview of the data, we will create a heatmap in which we look for the interactors with MSN2/4 and HSF1.

# TODO: create Heatmap $\to$ filter: flatfile for interactors $\to$ those in complex data for heatmap
```{r}

```

#### Observations of heatmap

# TODO: what can we see in the heatmap, what sticks out?

### Plot for foldchanges with DEseq2 

# TODO: do the plot
-> plot everything (facetwrap?) scatter to plot conditions against each other \newLine 

## Use DESeq2 

First of all, we installed the packages needed to use the tool.

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
```

Afterwards, we searched for some tutorials for how to use it. \newLine

For DEseq2 "non-normalized sequence read counts at either the gene or transcript level" (https://genviz.org/module-04-expression/0004/02/01/DifferentialExpression/) can be analyzed. Hence, we can use the package to analyze our complex_yeast_heatshock data. \newLine

Then we used the Help of R and looked for "DEseq2" to find relevant functions for our data. We then saw the function "DEseqDataSetFromMatrix()" which might be useful to read in our data-frame. In the next code chunk we will try some things with that.

# TODO: try out DEseq functions and analysis

$\to$ plot significant foldchanges , alpha _ 0.01 + do correction (multiple testing correction) \newLine

FOR THAT: look at HUBER chapter 8

```{r}

```

## Final Results in Cytoscape (PART 5)

# TODO: think of a graph to show here after the other analysis are done

## Discuss topics of chapter 13 in MSMB

### limitations of the outset

Since we didn't execute the experiment by ourselves, we don't have any idea about what limitations there might have been or what side effects could be there. Hence, we cannot take that into consideration but we could compare our findings to common findings on the internet/in papers.

### biased data

As mentioned before, we do not know about any biases that might appear. Nevertheless, the experiment was performed on yeast and the data was probably collected in a lab and it's not too hard to control the temperature with the needed tools. Also, the measurements were taken in different groups and not all on the same probe which also makes it better to compare the overall readcount per replicate and condition. Additionally, we do not know about the knockouts/Knockdowns which is why we are careful with our findings and look for plausibility of them.

### number of replicates

We only got 3 replicates of each condition which would definitely not be enough if we actually wanted to proof new findings. But it should be enough for us if we want to learn about the analysis of data and can look up other resources.

```{r}
# eg compare two conditions
mean1 <- mean(complex_yeast_heatshock$`25_0_Wildtype_r1`) + mean(complex_yeast_heatshock$`25_0_Wildtype_r2`) + mean(complex_yeast_heatshock$`25_0_Wildtype_r3`)
mean1 <- mean1/3
mean2 <- mean(complex_yeast_heatshock$`42_30_Double.KD/KO_r1`) + mean(complex_yeast_heatshock$`42_30_Double.KD/KO_r2`) + mean(complex_yeast_heatshock$`42_30_Double.KD/KO_r3`)
mean2 <- mean2/3
library("pwr")
str(pwr.t.test)
# two sample t test 
differenceMean = mean1/mean2
# calculates power we currently have
pwr.t.test(power = NULL, n=3, sig.level = 0.01, type = "two.sample", d = 2)

# calculates n needed for power of 0.8
pwr.t.test(power = 0.8, n = NULL,  sig.level = 0.01, type = "two.sample", d = 2)
```
As shown in the previous code chunk, the power when comparing two conditions of our data is only ca. 0.16 when we want to observe genes expressed twice or half as much. If we want a power over 0.8 to observe prominent changes we would need 8 samples for each condition.

### False discovery rate 

In order to control the false discovery rate, one should use multiple testing correction to find actually significant data. 

### Units vs fold changes

Since we often cannot use standardized units in biology, we have to be careful in comparing the data. As we use read counts, we should compare the variances in the different samples and measurements in order to make them comparable and to not have findings based on overall differences in the data due to the method of measuring etc. If an overall difference in the counts can be identified, one should use transformations to observe actual differentially expressed data and to not be misleaded. 




